main/autogen/autoloader.cc

```diff
@@ -388,31 +390,26 @@ void DefTreeBuilder::collapseSameFileDefs(const core::GlobalState &gs, const Aut
     }
 }
 
-void AutoloadWriter::writeAutoloads(const core::GlobalState &gs, const AutoloaderConfig &alCfg, const std::string &path,
-                                    const DefTree &root) {
-    UnorderedSet<string> toDelete; // Remove from this set as we write files
-    if (FileOps::exists(path)) {
-        vector<string> existingFiles = FileOps::listFilesInDir(path, {".rb"}, true, {}, {});
-        toDelete.insert(make_move_iterator(existingFiles.begin()), make_move_iterator(existingFiles.end()));
-    }
-    write(gs, alCfg, path, toDelete, root);
-    for (const auto &file : toDelete) {
-        FileOps::removeFile(file);
-    }
-}
-
-void AutoloadWriter::write(const core::GlobalState &gs, const AutoloaderConfig &alCfg, const std::string &path,
-                           UnorderedSet<std::string> &toDelete, const DefTree &node) {
+namespace {
+struct RenderAutoloadTask {
+    string filePath;
+    const DefTree &node;
+};
+
+// This function has two duties:
+// * It creates autoload rendering tasks which will occur in a later parallel phase.
+// * It creates subdirectories when needed, as they are required to write the autoloader output.
+void populateAutoloadTasksAndCreateDirectories(const core::GlobalState &gs, vector<RenderAutoloadTask> &tasks,
+                                               const AutoloaderConfig &alCfg, string_view path, const DefTree &node) {
     string name = node.root() ? "root" : node.name().show(gs);
     string filePath = join(path, fmt::format("{}.rb", name));
     if (!alCfg.packagedAutoloader || !node.root()) {
-        FileOps::writeIfDifferent(filePath, node.renderAutoloadSrc(gs, alCfg));
+        tasks.emplace_back(RenderAutoloadTask{filePath, node});
     }
-    toDelete.erase(filePath);
     if (!node.children.empty()) {
         auto subdir = join(path, node.root() ? "" : name);
-        if (!node.root() && !FileOps::dirExists(subdir)) {
-            FileOps::createDir(subdir);
+        if (!node.root()) {
+            FileOps::ensureDir(subdir);
```

Comment:
We still create dirs serially, though. Because we're traversing an autogen tree that corresponds directly with the file hierarchy, these `ensureDir` calls need to be ordered so that parent dirs are created prior to child dirs.